_drafts/swim.md

```diff
@@ -0,0 +1,441 @@
+    and its detection by *some* other node is at most
+    <code>T' * (1 / (1 - e^Qf))</code>, where <code>Qf</code> is the fraction of
+    non-faulty nodes in the cluster and <code>T'</code> is the protocol period
+    (in time units, should be at least three times the round-trip estimate) ;
+* **False positives**: the probability of direct or indirect probes being lost
+    is (gross approximation here, the real equation is a bit larger) <code>(1 - Delivery_%²)*(1 - Delivery_%⁴)^k</code>
+    — where <code>Delivery_%</code> is the probability of timely delivery of a
+    packet and <code>k</code> is the number of members used to perform indirect
+    probing). Notice how <code>k</code> is used as an exponent? It shows that
+    **increasing the number of nodes used for indirect probing will rapidly
+    decrease the probability of having false positives** ;
+* **Accuracy**: as the accuracy is directly correlated to the false positives,
+    we can calculate that with 95% of the packets successfully delivered and 3
+    nodes used for indirect probing, the protocol will have an accuracy of
+    99.9%. Again, this value increases as we increase <code>k</code> ;
+* **Network Message Load**: as each node only pings one target each protocol period,
+    we moved from an <code>O(N²)</code> to something in <code>O(N)</code>. We
+    roughly have two packets (<code>PING</code> and <code>ACK</code>) and an
+    optional indirect probe (hence the 4 extra packets per relay node)
+    <code>(4*k + 2)*N</code>.
+
+**Important:** the speed of failure detection and the probability of false
+positives don't depend on the number of nodes in the cluster (except
+asymptotically). Which means that the protocol will scale nicely, even on large
+clusters.
+
+## A more robust and efficient *SWIM*
+
+The version of the *SWIM* protocol we described so far could be qualified as «
+*basic SWIM* ». It works and showcases the main ideas of the protocol, but it
+suffers from a few issues and limitations:
+
+ * the dissemination component still relies on network multicasts. However,
+    network multicast primitives such as IP multicast etc., are only best-effort
+    and most of the time they aren't even enabled on production systems ;
+ * the failure detector doesn't handle well nodes that are perturbed for small
+    durations of time (overloaded host, unexpectedly long GC pause, …) ;
+ * the basic SWIM failure detection protocol guarantees eventual detection of
+    the failure of an arbitrary node. However, it gives no deterministic
+    guarantees on the time between failure of an arbitrary node and its
+    detection at another arbitrary member (in terms of the number of local
+    protocol rounds).
+
+Let's dig in these limitations to better understand them and see what can be
+done to go around.
+
+### Infection-style dissemination
+
+The basic protocol propagates membership updates (joins, leaves and failures)
```

Comment:
You have not described the mechanisms of joins and leaves, might be interesting here to add a link where they are explained